[3006.x] Update 3006 packaging to reduce permissions given to salt user for running salt-master

barneysowood opened the pull request:
### What does this PR do?

Updates Debian and RPM packages to improve isolation offered by running the `salt-master` processes under the `salt` user.

Implementation of running salt-master as a non-root user by default introduced in https://github.com/saltstack/salt/pull/64037, gives ownership of shared salt directories to the `salt` user to mitigate a number of issues with running the `salt-master` as a non-root user. This PR aims to reduce the scope of those permission grants to increase isolation of the `salt-master` processes running under the `salt` user - see https://github.com/saltstack/salt/issues/64193 for more detail.

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/64193

### Previous Behavior
The `salt` user that the `salt-master` process runs as was able to:

* Write to `/opt/saltstack/salt` modifying the salt install, including python, shared libs and python modules used by both the salt master and other salt process (eg `salt-api` and `salt-minion`) which run as root, compromising the isolation offered by running the `salt-master` as a non-root user.
* Read and write configs in `/etc/salt` for other salt daemons (that run as root) compromising the isolation offered by running as a non-root user.
* Reading private data from `/etc/salt/pki/minion` and `/var/cache/salt/minion` again compromising isolation.

### New Behavior

Debian and RPM packages will now:

* Keep ownership of `/opt/saltstack/salt` hierarchy as `root:root` to prevent modification by `salt-master` process
* Bytecompile python packages under `/opt/saltstack/salt/lib/` on install so that performance isn't compromised and the bytecompiled modules are owned by root
* Clean up *.pyc files and \_\_pycache__ dirs under `/opt/saltstack/salt/lib` on uninstall
* Under /etc/salt limit ownership to  /etc/salt/pki/master and /etc/salt/master.d
* Under /var/cache/salt and /var/run/salt only give ownership on master directories
* Under /var/log/salt, ensure /var/log/salt/master exists and give ownership of that. Also update logrotate config to create that with correct ownership and perms and install that on debian packages.

This should ensure that a compromised `salt-master` process cannot:

* Modify the salt installation used by it or other salt-daemons
* Modify the config of other salt daemons or tools
* Read pki or other private data *via the local filesystem* for a `salt-minion` running on the same host

### Merge requirements satisfied?

- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
No

barneysowood committed:
Update packages for python modules owned by root

Updates Debian and RPM packages so that /opt/saltstack/salt and the
python packages don't have to be owned by salt user.

It shouldn't be necessary for salt user, used to run salt-master to be
able to write/modify files in that directory hierarchy.

Add postinst scripts to call the python compileall module to create
byte-compiled python modules. This should preserve performance whilst
not requiring write access for salt user.

Also cleans up .pyc files and __pycache__ dirs on removal.

barneysowood committed:
Reduce perms for salt user on other salt dirs

Reduces the permissions granted to the salt user used to run the
salt-master:

* Under /etc/salt limit ownership to  /etc/salt/pki/master and
  /etc/salt/master.d
* Until #64219 is resolved also include /etc/salt/minion.d
* Under /var/cache/salt and /var/run/salt only give ownership on
  master directories
* Under /var/log/salt, ensure /var/log/salt/master exists and give
  ownership oof that. Also update logrotate config to create that with
  correct ownership and perms and install that on debian packages.

dwoz commented:
This is going to be affected by #64174

barneysowood commented:
As discussed in [#64174](https://github.com/saltstack/salt/pull/64174#issuecomment-1531660628), will wait for that to be merged then re-base this and integrate on top of those changes.
Also need to add some tests.

barneysowood commented:
Note that issues in https://github.com/saltstack/salt/issues/64219 will affect this.

barneysowood committed:
Update files to verify in pkg tests

Updates lists of files to check in package tests when starting master.
We now set the following as owned by salt:salt in master postinst

* /etc/salt/pki/master
* /etc/salt/master.d
* /var/log/salt/master
* /var/cache/salt/master
* /var/run/salt/master

barneysowood committed:
Create /var/run/salt/master

Creates /var/run/salt/master directory - if we don't create that
specifically the postinst script will fail

dwoz commented:
What is the `saltZ` user?

barneysowood commented:
> What is the `saltZ` user?

I'm getting a test failure that is throwing an error on chmod for the salt user not existing. That was just a hack to try and work out which chmod call was causing the issue. 

It'd really help if I could get the artifacts from test runs to get the pytest logs - they don't seem to be accessible from the workflows page - is there some other way to get access to them?

barneysowood committed:
Correct docstrings for salt user test_salt_user

Corrects docstrings for salt user tests - they were all the same.

barneysowood committed:
Move log creation and chown to posttrans

Moves log creation for /var/log/salt/master and /var/log/salt/cloud
and setting ownership to salt:salt to the %posttrans scriplets.
Whilst using %post work fine for fresh installs, upgrading means that
the previous package %postun removes those files. Using %posttrans
ensures the logs are created at the end of the full install/upgrade
transaction.

barneysowood committed:
Create empty log for salt-api

Creates empty log for salt-api owned by salt user in same way we do for
the master and salt-cloud

barneysowood committed:
Add tests for package directory and file ownership

Adds test that checks that files and directories created by the packages
that should be owned by salt:salt are owned by that user and that the
other files/directories created are owned by root:root.

barneysowood committed:
Remove group test for files

Removes group test for ownership on files by the salt user. Files that
are created by the salt-master process can be owned by salt:root, rather
than salt:salt and that's valid

barneysowood committed:
Handle creation of /var/log/salt/key

barneysowood committed:
Handle pytest-salt-factories permission changes

Handles permission changes caused by test suite running as root and
the creation of /var/cache/salt/master/.root_key.

Running the test suite as root means that /etc/salt/pki/master subdirs
get their ownership changed to root - clean that up in conftest.py.
No longer need to fix /var/log/salt as we handle the files in there
individually.

Adds exclusion for /var/cache/salt/master/.root_key as that gets created
by salt* cli tools running as root.

barneysowood committed:
Ensure salt-api service is enabled

Ensure salt-api service is enabled now we've added a postinst script

dmurphy18 commented:
@barneysowood Wondering if you have a timeframe for when this will come out of draft, itching to review, but don't want to, till it is out of draft.

barneysowood commented:
> @barneysowood Wondering if you have a timeframe for when this will come out of draft, itching to review, but don't want to, till it is out of draft.

@dmurphy18 - apologies, I've been out of action for a while. I'm planning to pick this up later this week, IIRC I've just got some issues with tests to fix. If I can get those sorted quickly I will do so, if not I may take it out of draft for review anyway whilst I work on those.

barneysowood committed:
Removing check on /etc/salt/minion.d

Removes check on /etc/salt/minion.d - it's causing issues in CI
that I can't reproduce locally and we'll deal with it in #64235
anyway.

barneysowood committed:
Remove some perms checks in conftest.py

Remove checks for perms on /var/log/salt/master and /var/run/salt/master
in the salt_master fixture as they may not existing during an upgrade
test and it's not critical to test them in the fixture.

Ch3LL commented:
@barneysowood this is still in draft mode. We hope to release the next bug fix really soon, mid August. Is this close to being ready?

barneysowood committed:
Add support for fixing old pkg perms

Adds support for fixing old package (3006.0/3006.1) perms that used the
salt user too widely.
Without doing this, tests don't pass for upgrades.

barneysowood commented:
> @barneysowood this is still in draft mode. We hope to release the next bug fix really soon, mid August. Is this close to being ready?

@Ch3LL - I think so. I'm hoping to have to one remaining issue that blocking tests passing fixed today. I'll let you know how that goes.

barneysowood commented:
re-run all

barneysowood commented:
Seeing a strange issue with tests for this. All tests now passing apart from "Test (upgrade, 3006.1)" for Debian 10/11 and Ubuntu 20.04/22.04 - see https://github.com/saltstack/salt/actions/runs/5801918905/job/15729830349?pr=64194 for an example.

What's failing specifically is `pkg/tests/integration/test_pip.py::test_pip_non_root` - 

```                   # Now, we should still not be able to install as non-root
                   ret = subprocess.run(
                       install_salt.binary_paths["pip"] + ["install", "pep8"],
                       preexec_fn=demote(test_account.uid, test_account.gid),
                       env=test_account.env,
                       stdout=subprocess.PIPE,
                       stderr=subprocess.PIPE,
                       check=False,
                       universal_newlines=True,
                   )
           >       assert ret.returncode != 0, ret.stderr
           E       AssertionError:
```

So that's testing that running `/opt/saltstack/salt/salt-pip install pep8` as a non-root user should fail. I can't recreate this when testing locally - taking the built artifacts and applying them to a VM with 3006.1 installed doesn't show the same behaviour.

Have also tested using the `ghcr.io/saltstack/salt-ci-containers/packaging:debian-11` container locally running `nox -e "test-pkgs-onedir" -- upgrade --prev-version 3006.1` and that doesn't show the failure.

Slightly stumped at this point - I suspect some interaction with pytest-salt-factories and the specific setup with CI, but that's really hard to debug.

@Ch3LL, @s0undt3ch or @dmurphy18 - any ideas on where to go next?

As this is the one remaining issue on this PR from my point of view, going to take it out of draft..

s0undt3ch commented on pkg/debian/salt-master.postinst:
```diff
@@ -1,6 +1,14 @@
+      touch /var/log/salt/key
+      chmod 640 /var/log/salt/key
+    fi
+    chown -R salt:salt /etc/salt/pki/master /etc/salt/master.d /etc/salt/minion.d /var/log/salt/master /var/log/salt/key /var/cache/salt/master /var/run/salt/master
```
```suggestion
    chown -R salt:salt /etc/salt/pki/master /etc/salt/master.d /var/log/salt/master /var/log/salt/key /var/cache/salt/master /var/run/salt/master
```
Should `/etc/salt/minion.d` really be above?

s0undt3ch commented on pkg/debian/salt-master.preinst:
```diff
@@ -0,0 +1,14 @@
+    [ -z "$SALT_USER" ] && SALT_USER=salt
+    [ -z "$SALT_NAME" ] && SALT_NAME="Salt"
+    [ -z "$SALT_GROUP" ] && SALT_GROUP=salt
+    PY_VER=$(/opt/saltstack/salt/bin/python3 -c "import sys; sys.stdout.write('{}.{}'.format(*sys.version_info)); sys.stdout.flush;")
```
```suggestion
    PY_VER=$(/opt/saltstack/salt/bin/python3 -c "import sys; sys.stdout.write('{}.{}'.format(*sys.version_info)); sys.stdout.flush();")
```

barneysowood replied:
Fixed

s0undt3ch commented on pkg/rpm/salt.spec:
```diff
@@ -406,8 +408,14 @@ usermod -c "%{_SALT_NAME}" \
+
+%pre master
+# Reset permissions to fix previous installs
+PY_VER=$(/opt/saltstack/salt/bin/python3 -c "import sys; sys.stdout.write('{}.{}'.format(*sys.version_info)); sys.stdout.flush;")
```
```suggestion
PY_VER=$(/opt/saltstack/salt/bin/python3 -c "import sys; sys.stdout.write('{}.{}'.format(*sys.version_info)); sys.stdout.flush();")
```

barneysowood replied:
Fixed

s0undt3ch commented on pkg/rpm/salt.spec:
```diff
@@ -480,6 +484,42 @@ ln -s -f /opt/saltstack/salt/salt-ssh %{_bindir}/salt-ssh
+  touch /var/log/salt/cloud
+  chmod 640 /var/log/salt/cloud
+fi
+chown -R %{_SALT_USER}:%{_SALT_GROUP} /etc/salt/cloud.deploy.d /var/log/salt/cloud /opt/saltstack/salt/lib/python3.10/site-packages/salt/cloud/deploy
```
```suggestion
chown -R %{_SALT_USER}:%{_SALT_GROUP} /etc/salt/cloud.deploy.d /var/log/salt/cloud /opt/saltstack/salt/lib/python3.10/site-packages/salt/cloud/deploy
```
Is there a way not to have `python3.10` hardcoded?

s0undt3ch commented on pkg/tests/integration/test_salt_user.py:
```diff
@@ -13,6 +14,52 @@
+    """
+    Paths created by package installs
+    """
+    paths = [
+        "/etc/salt",
+        "/var/cache/salt",
+        "/var/log/salt",
+        "/var/run/salt",
+        "/opt/saltstack/salt",
+    ]
+    return paths
```
```suggestion
    return [
        "/etc/salt",
        "/var/cache/salt",
        "/var/log/salt",
        "/var/run/salt",
        "/opt/saltstack/salt",
    ]
```

s0undt3ch commented on pkg/tests/integration/test_salt_user.py:
```diff
@@ -61,14 +108,54 @@ def test_salt_cloud_dirs(install_salt):
-        "/opt/saltstack/salt/lib/python{}.{}/site-packages/salt/cloud/deploy".format(
-            *sys.version_info
-        ),
+        "/opt/saltstack/salt/lib/python3.10/site-packages/salt/cloud/deploy",
```
```suggestion
        "/opt/saltstack/salt/lib/python3.10/site-packages/salt/cloud/deploy",
```
Don't hardcode the python version. Why did you need to change this line?

barneysowood replied:
Just accidentally stomped on that - reverted.

barneysowood replied:
I made the change in https://github.com/saltstack/salt/pull/64194/commits/1b411c7fb5cb4d07ddb24dae8b6c18953ea4d987 a couple of days before your comment.

s0undt3ch replied:
> I made the change in [1b411c7](https://github.com/saltstack/salt/commit/1b411c7fb5cb4d07ddb24dae8b6c18953ea4d987) a couple of days before your comment.

There's one more hardcoded reference, the one being commented on

barneysowood replied:
@s0undt3ch - apologies, you're correct. I've actually removed test_salt_cloud_dirs() as that is now covered by  test_pkg_paths()

barneysowood committed:
Fix call to sys.stdout.flush()

barneysowood committed:
Calculaate python version

barneysowood committed:
Return paths without setting var

barneysowood committed:
Revert to not hardcoding python version in cloud deploy path

barneysowood committed:
Remove seperate salt-cloud path tests

Removes seperate salt-cloud path tests as they are now covered by
test_pkg_paths

dmurphy18 reviewed (dismissed):
using perl for regex, hurts for python :)

Ch3LL commented:
@barneysowood these test failures look like they might be related. Note: We are trying to get everything merged in by this Wednesday for the next bug fix release.

barneysowood commented:
> @barneysowood these test failures look like they might be related. Note: We are trying to get everything merged in by this Wednesday for the next bug fix release.

@Ch3LL  - I've been trying to debug them, making some progress, but it's rather slow. When you say "these test failures look like they might be related" - what exactly do you mean? Related to what?

Ch3LL commented:
Related to the code changes in this PR, especially since I haven't seen this test failure in any other PRs. I will update the branch just to be 100% certain.

I have also asked @MKLeb to help you take a look at the failures, since we would like to get this in for the bug fix release. He said he can dive into it tomorrow to take a look.

barneysowood committed:
Add changelog

barneysowood commented:
> Related to the code changes in this PR, especially since I haven't seen this test failure in any other PRs. I will update the branch just to be 100% certain.
> 
> I have also asked @MKLeb to help you take a look at the failures, since we would like to get this in for the bug fix release. He said he can dive into it tomorrow to take a look.

@Ch3LL - ok thanks.

@MKLeb - would very much appreciate some help, pretty stumped with why `pkg/tests/integration/test_pip.py::test_pip_non_root` is failing on the 3006.2 upgrade. 

I created a branch with just some changes to that test to dump debug info and a PR to run the test suite.. results are weird.

Pulled the logs from this - https://github.com/saltstack/salt/actions/runs/5876536569/job/15938431199?pr=64947

The issue is that somehow running `salt-pip install` as the saltdev user works - it creates a dir in `/opt/saltstack/salt`.. I added some debugging to dump out `ls -l` for `/opt/saltstack`, `/opt/saltstack/salt` and `/opt/saltstack/salt/extras-3.10` before the run of `salt-pip install -h`, then before run of `salt-pip install pep8` and after that.

1. Output of `id saltdev`

```
uid=1002(saltdev) gid=1002(saltdev) groups=1002(saltdev)
```

2. Output of  `/bin/ls -l /opt/saltstack /opt/saltstack/salt /opt/saltstack/salt/extras-3.10` at start of test

```
/opt/saltstack:
total 4
drwxrwxr-x 6 root root 4096 Aug 16 10:42 salt

/opt/saltstack/salt:
total 92
drwxr-xr-x  2 root root  4096 Aug 16 10:40 bin
drwxr-xr-x 12 root root  4096 Aug 16 10:40 include
drwxr-xr-x  6 root root 20480 Aug 16 10:40 lib
-rwxr-xr-x  1 root root   295 Aug 16 08:37 salt
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-api
-rwxr-xr-x  1 root root   295 Aug 16 08:37 salt-call
-rwxr-xr-x  1 root root   297 Aug 16 08:37 salt-cloud
-rwxr-xr-x  1 root root   291 Aug 16 08:37 salt-cp
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-key
-rwxr-xr-x  1 root root   299 Aug 16 08:37 salt-master
-rwxr-xr-x  1 root root   299 Aug 16 08:37 salt-minion
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-pip
-rwxr-xr-x  1 root root   297 Aug 16 08:37 salt-proxy
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-run
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-ssh
-rwxr-xr-x  1 root root   299 Aug 16 08:37 salt-syndic
drwxr-xr-x  3 root root  4096 Aug 16 10:39 share
-rwxr-xr-x  1 root root   293 Aug 16 08:37 spm
-rwxr-xr-x  1 root root   282 Aug 16 08:37 wheel

 /bin/ls: cannot access '/opt/saltstack/salt/extras-3.10': No such file or directory
```

3. Output of  `/bin/ls -l /opt/saltstack /opt/saltstack/salt /opt/saltstack/salt/extras-3.10` after `salt-pip install -h` but before `salt-pip install pep8`

```
/opt/saltstack:
total 4
drwxrwxr-x 6 root root 4096 Aug 16 10:42 salt

/opt/saltstack/salt:
total 92
drwxr-xr-x  2 root root  4096 Aug 16 10:40 bin
drwxr-xr-x 12 root root  4096 Aug 16 10:40 include
drwxr-xr-x  6 root root 20480 Aug 16 10:40 lib
-rwxr-xr-x  1 root root   295 Aug 16 08:37 salt
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-api
-rwxr-xr-x  1 root root   295 Aug 16 08:37 salt-call
-rwxr-xr-x  1 root root   297 Aug 16 08:37 salt-cloud
-rwxr-xr-x  1 root root   291 Aug 16 08:37 salt-cp
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-key
-rwxr-xr-x  1 root root   299 Aug 16 08:37 salt-master
-rwxr-xr-x  1 root root   299 Aug 16 08:37 salt-minion
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-pip
-rwxr-xr-x  1 root root   297 Aug 16 08:37 salt-proxy
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-run
-rwxr-xr-x  1 root root   293 Aug 16 08:37 salt-ssh
-rwxr-xr-x  1 root root   299 Aug 16 08:37 salt-syndic
drwxr-xr-x  3 root root  4096 Aug 16 10:39 share
-rwxr-xr-x  1 root root   293 Aug 16 08:37 spm
-rwxr-xr-x  1 root root   282 Aug 16 08:37 wheel

/bin/ls: cannot access '/opt/saltstack/salt/extras-3.10

```

4. Output of  `/bin/ls -l /opt/saltstack /opt/saltstack/salt /opt/saltstack/salt/extras-3.10` after `salt-pip install pep8`

```
/opt/saltstack:
total 4
drwxrwxr-x 7 root root 4096 Aug 16 10:42 salt

/opt/saltstack/salt:
total 96
drwxr-xr-x  2 root    root     4096 Aug 16 10:40 bin
drwxr-xr-x  5 saltdev saltdev  4096 Aug 16 10:42 extras-3.10
drwxr-xr-x 12 root    root     4096 Aug 16 10:40 include
drwxr-xr-x  6 root    root    20480 Aug 16 10:40 lib
-rwxr-xr-x  1 root    root      295 Aug 16 08:37 salt
-rwxr-xr-x  1 root    root      293 Aug 16 08:37 salt-api
-rwxr-xr-x  1 root    root      295 Aug 16 08:37 salt-call
-rwxr-xr-x  1 root    root      297 Aug 16 08:37 salt-cloud
-rwxr-xr-x  1 root    root      291 Aug 16 08:37 salt-cp
-rwxr-xr-x  1 root    root      293 Aug 16 08:37 salt-key
-rwxr-xr-x  1 root    root      299 Aug 16 08:37 salt-master
-rwxr-xr-x  1 root    root      299 Aug 16 08:37 salt-minion
-rwxr-xr-x  1 root    root      293 Aug 16 08:37 salt-pip
-rwxr-xr-x  1 root    root      297 Aug 16 08:37 salt-proxy
-rwxr-xr-x  1 root    root      293 Aug 16 08:37 salt-run
-rwxr-xr-x  1 root    root      293 Aug 16 08:37 salt-ssh
-rwxr-xr-x  1 root    root      299 Aug 16 08:37 salt-syndic
drwxr-xr-x  3 root    root     4096 Aug 16 10:39 share
-rwxr-xr-x  1 root    root      293 Aug 16 08:37 spm
-rwxr-xr-x  1 root    root      282 Aug 16 08:37 wheel

/opt/saltstack/salt/extras-3.10:
total 96
drwxr-xr-x 2 saltdev saltdev  4096 Aug 16 10:42 bin
drwxr-xr-x 2 saltdev saltdev  4096 Aug 16 10:42 pep8-1.7.1.dist-info
-rw-r--r-- 1 saltdev saltdev 82878 Aug 16 10:42 pep8.py
drwxr-xr-x 2 saltdev saltdev  4096 Aug 16 10:42 __pycache__

```

[full log](https://github.com/saltstack/salt/files/12358635/runtests-upgrade.log)

So, somehow, despite `/opt/saltstack/salt` being `0775 root:root`, the saltdev user is able to create a `extras-3.10` in it. Maybe I'm missing something obvious here - I have been staring at it a long time.

Ch3LL commented:
@barneysowood looks like there is a merge conflict that needs to be resolved here now. I'm guessing it was caused by this: https://github.com/saltstack/salt/pull/64378

barneysowood commented:
> @barneysowood looks like there is a merge conflict that needs to be resolved here now. I'm guessing it was caused by this: #64378

Resolved that.

barneysowood committed:
Don't change ownership of /etc/salt/minion.d

No longer changes ownership of /etc/salt/minion.d to salt user for the
salt-master. Requires #64219 to be resolved.

barneysowood commented:
Created #65001 - which can be merged once this and #64235 have been merged to 3006.x

dwoz commented:
Have we tested upgrading from a 3005.x package?

dwoz commented:
> Have we tested upgrading from a 3005.x package?

@MKLeb Let me know we have tests that test the upgrades from 3005.x.

MKLeb commented on pkg/debian/salt-common.preinst:
```diff
@@ -31,11 +31,5 @@ case "$1" in
             -s $SALT_SHELL  \
             -g $SALT_GROUP  \
              $SALT_USER
-    # 5. adjust file and directory permissions
```
Hi @barneysowood, are you able to elaborate on why you are removing this piece?

barneysowood replied:
Hi @MKLeb - because the `salt` owner that the salt-master process runs as shouldn't be able to write to /opt/saltstack/salt - this would allow it modify the whole salt install...

barneysowood commented:
> > Have we tested upgrading from a 3005.x package?
> 
> @MKLeb Let me know we have tests that test the upgrades from 3005.x.

We do and I've also manually tested that on Ubuntu and CentOS.

Commit:
Fix `test_pip_non_root`
